Multiblank Transducer

hainan-xv opened the pull request:
# What does this PR do ?

Add support for multi-blank transducers described in https://arxiv.org/pdf/2211.03541

**Collection**:

ASR

# Changelog 
- implementation of multi-blank both with numba kernel code as well as pytorch.
- add relevant test code.
- add example .yaml file for multi-blank ASR model configuration.

# Usage

Invoke transducer model training with examples/asr/conf/conformer/conformer_multiblank_transducer_bpe.yaml file. This will train a multi-blank model with big blank durations [2,4,8], and sigma=0.05. 

To run multi-blank with custom configurations, e.g. using [2,3,5,8], change the options by,

```
model.loss.multiblank_rnnt_kwargs.big_blank_durations=[2,3,5,8] # used for loss computation in training.
model.loss.multiblank_rnnt_kwargs.sigma=0.05 # 0.05 is the recommended value.
model.decoding.big_blank_durations=[2,3,5,8] # used for decoding. The length of the list should be consistent with above.
model.joint.num_extra_outputs=4 # this needs to be consistent with the duration list you set
```

You could set model.decoding.big_blank_durations to have values smaller than what's used in training (model.loss.multiblank_rnnt_kwargs.big_blank_duration), e.g. [1,1,1,1]. This makes decoding slower but usually gives better accuracy.

To run inference, add option rnnt_decoding.big_blank_durations when calling NeMo/examples/asr/transcribe_speech.py, to specify the big blank durations, e.g.

```
python NeMo/examples/asr/transcribe_speech.py \
      model_path=your_big_blank_model.nemo \
      dataset_manifest=your_data.json  \
      ++rnnt_decoding.big_blank_durations=[2,3,5,8] \
      output_filename=your_output_file
```


# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

artbataev commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,806 @@ def _get_initial_states(self, batchsize):
+        decoder_model: rnnt_abstract.AbstractRNNTDecoder,
+        joint_model: rnnt_abstract.AbstractRNNTJoint,
+        blank_index: int,
+        big_blank_durations: int,
```
is it `List[int]`?

hainan-xv replied:
Done.

artbataev reviewed (commented):
The loss itself looks good, nice clean code! 
Do you have plans to add small (unit or end-to-end) tests for multiblank decoding?

hainan-xv commented on nemo/collections/asr/losses/rnnt_pytorch.py:
```diff
@@ -110,3 +110,126 @@ def compute_forward_prob(self, acts, labels, act_lens, label_lens):
+    def compute_forward_prob(self, acts, labels, act_lens, label_lens):
+        B, T, U, _ = acts.shape
+
+        log_alpha = torch.zeros(B, T, U)
```
Good idea. Done!

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -1046,7 +1073,7 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
             A list of decoded tokens.
         """
-        token_list = [self.labels_map[c] for c in tokens if c != self.blank_id]
+        token_list = [self.labels_map[c] for c in tokens if c < self.blank_id]
```
Actually here in line 1035 (__init__ function of class RNNTDecoding) it has blank_id = len(vocabulary) and passed this to the parent class AbstractRNNTDecoding. So probably it's fine here?

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -438,7 +465,7 @@ def decode_hypothesis(self, hypotheses_list: List[Hypothesis]) -> List[Union[Hyp
             # RNN-T sample level is already preprocessed by implicit RNNT decoding
             # Simply remove any blank tokens
-            prediction = [p for p in prediction if p != self.blank_id]
+            prediction = [p for p in prediction if p < self.blank_id]
```
Yeah using a set is probably a good idea I should've started with. Now though, a lot of the code, including the kernel code, assumes blank is the last symbol. I guess let me just add some checks in the code for the blank symbol: if we're using multi-blank RNNTs, blank id must be after non-blank ids; I'll make sure that for standard RNN-Ts there's no such limitation.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+    xlen: torch.Tensor,
+    ylen: torch.Tensor,
+    mlabels: torch.Tensor,
+    minibatch: int,
```
Hmm you're right. I never noticed that... I was copying the function signatures of the original RNN-T functions in this file, and this variable was not used in those functions either.

@titu1994 shall we remove this variable from all function signatures in this file?

titu1994 replied:
Keep the arguments. I put them all there because i wanted the signatures to be similar for all kernels, and i expect future kernels may eventually need these parameters. In such a case, i want to avoid having different signature for different kernels v

hainan-xv commented:
> Do you have plans to add small (unit or end-to-end) tests for multiblank decoding?

Sure I can add some tests for the decoding. But I can't seem to find any existing test templates for decoding? What kind of test do you have in mind @artbataev ?

artbataev commented:
> > Do you have plans to add small (unit or end-to-end) tests for multiblank decoding?
> 
> Sure I can add some tests for the decoding. But I can't seem to find any existing test templates for decoding? What kind of test do you have in mind @artbataev ?

I just thought about potential issues if someone touches the decoding code. Unfortunately, I don't see any existing tests for decoding. 
It would be great to have some, but not sure it's necessary.

I can propose to save logits and etalon decoding result for some not-very-long utterance (e.g., from LibriSpeech) and use it to test the decoder. Or to generate logits on the fly with obvious greedy decoding results (but this may be more tricky).

titu1994 commented:
I think I recently added decoding tests and they add usually quite necessary. 

https://github.com/NVIDIA/NeMo/blob/main/tests/collections/asr/test_asr_rnnt_encdec_model.py#L292

Tests below this are for decoding layer. We should have similar mock test for this loss. 

We can generate obvious greedy decoding logits with small TxUxV dim to simulate results, or generate randomly to check if they execute safely. I would prefer simulated result though.

titu1994 commented:
We also test loss on hardcoded activations, no reason we can't do it for this loss + decoding too. https://github.com/NVIDIA/NeMo/blob/main/tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py#L73

hainan-xv commented:
> We also test loss on hardcoded activations, no reason we can't do it for this loss + decoding too. https://github.com/NVIDIA/NeMo/blob/main/tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py#L73

@titu1994 I already included the test for the loss values and gradients. I will prepare some tests for decoding. Thanks!

hainan-xv commented:
Added test for multi-blank RNN-T decoding.

hainan-xv commented:
@artbataev @fayejf I have addressed the issues raised. Any more comments?

artbataev commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@
+            for hyp in hypotheses:
+                hyp.alignments = [[]]
+        else:
+            alignments = None
```
Seems to be really unused. @hainan-xv, is this necessary?

titu1994 replied:
The code ? Or this alignment? Alignment is used, this fallback on the other hand, not so much. If you want, you can ignore this codepath but let's try to have it working if possible / easy

hainan-xv replied:
I am actuallly confused about this too. I copied this from the original decoding method for RNN-Ts and it has this too. Could it be a bug and it was meant to be `hyp.alignments = None`?

titu1994 replied:
Hyp.alignments is None by default. This is just local var that is unused. We can update it to hyp.alignments though

artbataev reviewed (dismissed):
@hainan-xv looks good. Thanks for fixing the issues! 
Looking forward to new cool models! :)

fayejf commented on Dockerfile:
```diff
@@ -14,7 +14,7 @@
 # limitations under the License.
 
-ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:22.09-py3
+ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:22.04-py3
```
revert this please

titu1994 commented on nemo/collections/asr/modules/rnnt.py:
```diff
@@ -1174,6 +1174,7 @@ def __init__(
         self,
         jointnet: Dict[str, Any],
         num_classes: int,
+        num_extra_outputs: int = 0,
```
Same with the "num_classez_with_blank"

hainan-xv replied:
Done

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                        alphas[offset + t * maxU + u] = rnnt_helper.log_sum_exp(
+                            alphas[offset + t * maxU + u],
+                            alphas[offset + (t - big_blank_duration[i]) * maxU + u]
+                            + logp(
```
Imo it really helps me and other folks when we need to debug or extend. The original c++ code had exactly 0 comments for The kernels, absolute nightmare fuel

hainan-xv replied:
Unfortunately I didn't include the gradient derivation in the paper... I guess I can prepare some write-up detailing the maths behind the method later...

hainan-xv replied:
I added equations in the comments for each place I changed in the code.

titu1994 replied:
Thanks !

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                no_emit = (
+                    alphas[offset + (t - 1) * maxU + u]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t - 1, u, blank_)
+                    - sigma
```
Note the sigma subtraction inline is for multiblank

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                betas[offset + t * maxU + U - 1] = (
+                    betas[offset + (t + 1) * maxU + U - 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, U - 1, blank_)
+                    - sigma
```
Sigma expln inline

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                            betas[offset + (t + big_blank_duration[i]) * maxU + U - 1]
+                            + logp(denom, acts, maxT, maxU, alphabet_size, b, t, U - 1, blank_ - 1 - i)
+                            - sigma,
+                        )
```
Here too

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                    if t + big_blank_duration[i] < T:
+                        betas[offset + t * maxU + U - 1] = rnnt_helper.log_sum_exp(
+                            betas[offset + t * maxU + U - 1],
+                            betas[offset + (t + big_blank_duration[i]) * maxU + U - 1]
```
Eqn if possible

hainan-xv replied:
Done,

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                betas[offset + (T - 1) * maxU + u] = (
+                    betas[offset + (T - 1) * maxU + u + 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, T - 1, u, labels[u])
+                    - sigma
```
Sigma inline

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                emit = (
+                    betas[offset + t * maxU + u + 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, u, labels[u])
+                    - sigma
```
Two able

hainan-xv replied:
What's two able?

titu1994 replied:
Hah i meant refer to two above. Phone is useless for reviews sometimes.

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                            betas[offset + (t + big_blank_duration[i]) * maxU + u]
+                            + logp(denom, acts, maxT, maxU, alphabet_size, b, t, u, blank_ - 1 - i)
+                            - sigma
+                        )
```
Eqn and sigma

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                grads[col * alphabet_size + idx] = g
+
+            # update internal index through the thread_buffer;
+            # until idx < V + 1, such that entire vocabulary has been updated.
```
Just a note, incredible work on this. Nearly nothing to change other than docstrings

hainan-xv replied:
Ha thanks.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@
+            confidence_method_cfg=confidence_method_cfg,
+        )
+        self.big_blank_durations = big_blank_durations
+        self._SOS = blank_index - len(big_blank_durations)
```
SOS should be the original blank no? At index -1?

hainan-xv replied:
The issue is this is used in an embedding look-up for the decoder at begin of sentences for generating the initial decoder state. If we make it index -1, then we will have to allocate embeddings for those longer blanks which will be never used. (Aside from begin-of-sentence, the blank symbosl never took part in the decoder computation).

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
         return input_states
 
 
+class GreedyMultiblankRNNTInfer(_GreedyRNNTInfer):
```
Hmm how about subclassing the impl of GreedRNNTinfer instead of the abstract class? Forward can be avoided redoing then and just override _greedy_decode?

titu1994 replied:
Well it's just an option, this is fine too

hainan-xv replied:
Good suggestion. Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+        big_blank_duration = 1
+
+        for time_idx in range(out_len):
+            if blank_optimization and big_blank_duration > 1:
```
Add comment about what's being done here. Yes i know it's obvious but still

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+                v, k = logp.max(0)
+                k = k.item()  # K is the label at timestep t_s in inner loop, s >= 0.
+
+                if k >= self._blank_index - len(self.big_blank_durations) and k < self._blank_index:
```
Add comment about step being done here. Essentially every line should be documented for something as complicated as RNNT decoding v

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+        return hypothesis
+
+
+class GreedyBatchedMultiblankRNNTInfer(_GreedyRNNTInfer):
```
Consider subclassing GreedyBatchedRNNTInfer, though this is fine too v

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@
+            self._greedy_decode = self._greedy_decode_blank_as_pad
+        else:
+            self._greedy_decode = self._greedy_decode_masked
+        self._SOS = blank_index - len(big_blank_durations)
```
Same comment as beforez shouldn't this be last token ?

hainan-xv replied:
Refer to my answer to your other comment.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+        encoder_output: torch.Tensor,
+        encoded_lengths: torch.Tensor,
+        partial_hypotheses: Optional[List[rnnt_utils.Hypothesis]] = None,
+    ):
```
Probably could be dropped if subclasses

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+            # Get max sequence length
+            max_out_len = out_len.max()
+
+            big_blank_masks = [torch.full([batchsize], fill_value=0, dtype=torch.bool, device=device)] * len(
```
Add comment to every line executed for RNNT infer

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+            big_blank_duration = 1
+
+            for time_idx in range(max_out_len):
+                if blank_optimization and big_blank_duration > 1:
```
Here

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:
```diff
@@ -1446,6 +1446,809 @@ def _get_initial_states(self, batchsize):
+                    if blank_mask.all():
+                        not_blank = False
+
+                        for i in range(len(big_blank_masks) + 1):
```
Here

hainan-xv replied:
Done.

titu1994 reviewed (changes requested):
Incredible work here, without a doubt the easiest pr I've reviewed despite 2k lines. 

Minor comments here and there, mostly inline docstrings, maybe some concerning _SOS behavior in the RNNT decoding classes. 

Kernels look perfect, need a bit more verbose docstrings (future proofing)'

hainan-xv commented on README.rst:
```diff
@@ -65,6 +65,7 @@ Key Features
     * `Automatic Speech Recognition (ASR) <https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/intro.html>`_
         * Supported models: Jasper, QuartzNet, CitriNet, Conformer-CTC, Conformer-Transducer, Squeezeformer-CTC, Squeezeformer-Transducer, ContextNet, LSTM-Transducer (RNNT), LSTM-CTC, ...
         * Supports CTC and Transducer/RNNT losses/decoders
+            * NeMo Original `Multi-blank Transducers <https://arxiv.org/abs/2211.03541>`_: `Example Configuration File <https://github.com/NVIDIA/NeMo/tree/stable/examples/asr/conf/conformer/conformer_multiblank_transducer_bpe.yaml>`_
```
What do you suggest that I have here then?

hainan-xv commented on examples/asr/conf/conformer/conformer_multiblank_transducer_bpe.yaml:
```diff
@@ -0,0 +1,262 @@
+      tsd_max_sym_exp: 50  # for Time Synchronous Decoding
+      alsd_max_target_len: 2.0  # for Alignment-Length Synchronous Decoding
+
+  loss:
```
OK

hainan-xv commented on nemo/collections/asr/losses/rnnt.py:
```diff
@@ -311,7 +348,7 @@ def forward(self, log_probs, targets, input_lengths, target_lengths):
         # Transcript: [B, L] -> [B, L']; If L' < L
         if targets.shape[1] != max_targets_len:
-            targets = targets.narrow(dim=1, start=0, length=max_targets_len)
+            targets = targets.narrow(dim=1, start=0, length=max_targets_len).contiguous()
```
Ha, I think in practice this function is rarely called. This is called in my tests and thus this bug was caught.

titu1994 replied:
Perfect thanks !

hainan-xv commented on nemo/collections/asr/losses/rnnt_pytorch.py:
```diff
@@ -110,3 +110,125 @@ def compute_forward_prob(self, acts, labels, act_lens, label_lens):
+                    else:
+                        # here both t and u are > 0, this state is reachable
+                        # with two possibilities: (t - d, u) with emission of
+                        # blank with duration d, or (t, u - 1) with a label emission.
```
Done.

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -437,8 +467,14 @@ def decode_hypothesis(self, hypotheses_list: List[Hypothesis]) -> List[Union[Hyp
+                    num_extra_outputs += len(self.big_blank_durations)
+                prediction = [p for p in prediction if p < self.blank_id - num_extra_outputs]
+            else:
+                prediction = [p for p in prediction if p != self.blank_id]
```
OK.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py:
```diff
@@ -91,6 +91,72 @@ def backward(ctx, grad_output):
+        if is_cuda:
+            loss_func = rnnt.multiblank_rnnt_loss_gpu
+        else:
+            exit(-1)
```
Done.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py:
```diff
@@ -168,6 +282,77 @@ def forward(self, acts, labels, act_lens, label_lens):
+    def __init__(
+        self,
+        blank,
+        big_blank_durations=[],
```
Oh right. Actually it's probably a bad idea to make this param optional. I am making this mandatory this this is Multiblank class afterall.

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -195,13 +195,17 @@ def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
         self.word_seperator = self.cfg.get('word_seperator', ' ')
 
+        if self.big_blank_durations is not None:
+            assert blank_id != 0, 'blank_id must equal len(vocabs) for multi-blank RNN-T models'
```
Done.

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -240,32 +244,58 @@ def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
-                preserve_frame_confidence=self.preserve_frame_confidence,
-                confidence_method_cfg=self.confidence_method_cfg,
-            )
+            if self.big_blank_durations is not None:
```
Done.

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -240,32 +244,58 @@ def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
-                confidence_method_cfg=self.confidence_method_cfg,
-            )
+            if self.big_blank_durations is not None:
+                self.decoding = greedy_decode.GreedyBatchedMultiblankRNNTInfer(
```
Done

hainan-xv commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -1005,8 +1041,11 @@ class RNNTDecoding(AbstractRNNTDecoding):
+            len(vocabulary) + joint.num_extra_outputs
+        )  # we need to ensure blank is the last token in the vocab. This is needed for multi-blank RNN-T models.
         self.labels_map = dict([(i, vocabulary[i]) for i in range(len(vocabulary))])
+        self.num_extra_outputs = joint.num_extra_outputs
```
Done.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                emit = (
+                    alphas[offset + t * maxU + u - 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, u - 1, labels[u - 1])
+                    - sigma
```
Added comments that explains this.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                    if t >= big_blank_duration[i]:
+                        big_blank_no_emit = (
+                            alphas[offset + (t - big_blank_duration[i]) * maxU + u]
+                            + logp(
```
Done

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+        loglike = (
+            alphas[offset + (T - 1) * maxU + U - 1]
+            + logp(denom, acts, maxT, maxU, alphabet_size, b, T - 1, U - 1, blank_)
+            - sigma
```
Explained.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+                big_blank_loglike = (
+                    alphas[offset + (T - big_blank_duration[i]) * maxU + U - 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, T - big_blank_duration[i], U - 1, blank_ - 1 - i)
+                    - sigma
```
Explained.

hainan-xv commented on nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py:
```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+        for i in range(num_big_blanks):
+            if T >= big_blank_duration[i] and big_blank_duration[i] == 1:
+                betas[offset + (T - 1) * maxU + U - 1] = rnnt_helper.log_sum_exp(
+                    betas[offset + (T - 1) * maxU + U - 1],
```
Actually I noticed this for loop of the code is from an older version and now is never called, thus not needed any more.

hainan-xv commented:
Addressed comments by @titu1994 and then did a rebase and squashed all comments. Some notable changes:
1. for greedy decoding, change the parent classes of the added classes to standard transducer decoding classes instead of the abstract class, so that forward() functions are not needed in those classes.
2. added detailed comments for decoding algorithm for multi-blank RNN-Ts.
3. added detailed comments for the multi-blank kernel code by reference equations.
4. use RaiseError() instead of exit() for a number of places where assertions fail.

What I am not sure of:
1. value of SOS: I set it to blank_id - len(big_blanks) because SOS is used in the decoder to generate hidden states for the begin-of-sentence, and this involves an embedding lookup in the internal computation of the decoder. Note decoder embedding is initialized as a tensor of size [non_blank_vocab_size + 1, emb_size], and thus using blank_id would cause an out-of-bound issue. If we initialize the tensor to [non_blank_vocab_size + 1 + num_big_blanks], then the added embeddings for the big blanks would never be used or updated.
2. `alignments = None` in nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py. I also find it weird but it is copied from the base class. My guess is it is a bug and it is supposed to be `hyp.alignments = None`?

titu1994 commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -191,17 +191,23 @@ class AbstractRNNTDecoding(ConfidenceMixin):
     """
 
-    def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
+    def __init__(self, decoding_cfg, decoder, joint, blank_id: int, num_extra_outputs: int = 0):
```
Instead of taking it as an argument, do joint.num_extra_outputs, and AbstractRNNTJoint will now have two properties - `num_classes_with_blank` and `num_extra_outputs` that will both return ints. By default the impl of these two properties is to raise ValueError().

Subclasses must implement these methods, calculating their vocab size and their num_extra_outputs. 

The decoding framework can then reliably access these values from the joint itself rather than accept such values via its constructor. The Decoding framework's constructor should only contain a few elements - the config, the decoder module, joint module and the blank id.

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -1005,10 +1043,18 @@ class RNNTDecoding(AbstractRNNTDecoding):
+            decoder=decoder,
+            joint=joint,
+            blank_id=blank_id,
+            num_extra_outputs=joint.num_extra_outputs,
```
Remove

hainan-xv replied:
Done.

titu1994 commented on nemo/collections/asr/metrics/rnnt_wer.py:
```diff
@@ -191,17 +191,23 @@ class AbstractRNNTDecoding(ConfidenceMixin):
         super(AbstractRNNTDecoding, self).__init__()
         self.cfg = decoding_cfg
         self.blank_id = blank_id
+        self.num_extra_outputs = num_extra_outputs
```
This line becomes self.joint.num_extra_outputs then.

hainan-xv replied:
Done.

titu1994 reviewed (commented):
Nearly perfect. Minor refactor to clean up decoding framework and make the AbstractRNNTJoint a bit more self contained/

titu1994 reviewed (approved):
Looks great, ready to merge

fayejf reviewed (approved):
LGTM!

fayejf reviewed (approved):
approve again

titu1994 commented:
Amazing work again, @hainan-xv  !
